src/main/scala/se/lth/immun/TargetDeisotoper.scala

```diff
@@ -1,57 +1,162 @@
+			
+      val batchTargetList = targets.sortBy(_.mz).grouped(targetBatchSize).toList
+      for (i <- 0 until batchTargetList.length)
+			  toProcess += i -> batchTargetList(i)
+			
+			if (params.verbose)
+  			println("created " + batchTargetList.length + " target deisotoping batches")
+			processIfFree
 		
-		println("=== IN TARGETED MODE - BEWARE!!! ===")
-		val hillsByMz = hills.sortBy(_.total.centerMz)
+		case Deisotoped(batchId, isotopes) =>      
+		  completedPatterns += isotopes
+		  beingProcessed -= batchId
+		  if (params.verbose)
+  		  println("batch deisotoping finished " + batchId + ", found " + isotopes.length + " unique isotope patterns")	
+		  
+		  if (toProcess.isEmpty && beingProcessed.isEmpty) {
+				if (params.verbose)
+				  println("deisotoping finished")
+				
+				val isotopesAll = completedPatterns.flatten.sortBy(ip => (ip.hills.head.total.centerMz, ip.z, ip.apexHill.accurateApexRt(specTime)))
+				if (params.verbose)
+				  println("sorting finished, " + isotopesAll.length + " isotope patterns found")
+				
+				def filterUnique(isotopesSorted:Seq[IsotopePattern]) = 
+		      isotopesSorted.foldRight(List.empty[IsotopePattern]){ 
+            case (a, b) => 
+              if (!b.isEmpty && b(0).hills.head.total.centerMz == a.hills.head.total.centerMz && b(0).z == a.z && b(0).apexHill.accurateApexRt(specTime) == a.apexHill.accurateApexRt(specTime)) 
+                b 
+              else 
+                a :: b 
+          }
+				val uniqueIsotopes = filterUnique(isotopesAll)
+				
+        if (params.verbose)
+				  println("filtered unique isotope patterns, " + uniqueIsotopes.length + " isotope patterns remaining")
+				
+				customer ! TargetDeisotopingComplete(Nil, uniqueIsotopes.toSeq)
+				context.stop(self)
+			} else
+			  processIfFree
+						
+	}
+	
+	
+	def processIfFree = {
+		while (toProcess.nonEmpty && beingProcessed.size < params.concurrency) {
+			val (batchId, batchTargets) = toProcess.dequeue
+			beingProcessed += batchId
+			val a = context.actorOf(Props(new TargetBatchDeisotoper(params)))
+  		a ! BatchDeisotope(batchId, hillsByMz, batchTargets, hillsMzs, hillsRts, hillsScanIndices, specTime)
+		}
+	}
+}
+
+class TargetBatchDeisotoper(val params:DinosaurParams) extends Actor {
+  
+  import Cluster._
+  
+  implicit val p = params
+  
+  def receive = {
 		
-		if (params.verbose)
-			println("deisotoping based on targets...")
+		case BatchDeisotope(batchId, hillsByMz, batchTargets, hillsMzs, hillsRts, hillsScanIndices, specTime) =>
+		  val (_, isotopes) = deisotope(hillsByMz, batchTargets, hillsMzs, hillsRts, hillsScanIndices)
+		  	  
+		  val uniqueIsotopes = filterUnique(isotopes.sortBy(ip => (ip.hills.head.total.centerMz, ip.z, ip.apexHill.accurateApexRt(specTime))), specTime)
+		  				  
+		  sender ! Deisotoped(batchId, uniqueIsotopes)
+		  context.stop(self)
+  }
+  
+  def filterUnique(isotopes:Seq[IsotopePattern], specTime:Seq[Double]) = 
```

Comment:
This method seems to be copy-pasted from above, and could be a shared static method instead.

Reply:
Agreed